File: backend/services/leave_service.py
from datetime import datetime, date
from typing import Optional
from firebase_admin_init import get_db
from models import SHORT_LEAVE_MAX_DAYS, SHORT_LEAVE_NOTICE_DAYS, LONG_LEAVE_NOTICE_DAYS


def _parse_date(date_str: str) -> date:
    return datetime.strptime(date_str, "%Y-%m-%d").date()


def days_between(start: str, end: str) -> int:
    return (_parse_date(end) - _parse_date(start)).days + 1


def days_from_today(date_str: str) -> int:
    return (_parse_date(date_str) - date.today()).days


def validate_leave(start_date: str, end_date: str) -> Optional[str]:
    start = _parse_date(start_date)
    end = _parse_date(end_date)

    if end < start:
        return "תאריך סיום חייב להיות אחרי תאריך התחלה"

    duration = days_between(start_date, end_date)
    days_ahead = days_from_today(start_date)

    if duration <= SHORT_LEAVE_MAX_DAYS:
        if days_ahead < SHORT_LEAVE_NOTICE_DAYS:
            return f"חופשה קצרה (1-2 ימים) דורשת {SHORT_LEAVE_NOTICE_DAYS} ימי הודעה מראש"
    else:
        if days_ahead < LONG_LEAVE_NOTICE_DAYS:
            remaining = LONG_LEAVE_NOTICE_DAYS - days_ahead
            return f"חופשה ארוכה (3+ ימים) דורשת {LONG_LEAVE_NOTICE_DAYS} ימי הודעה מראש (חסרים {remaining} ימים)"

    return None


def request_leave(
    employee_id: str,
    employee_name: str,
    start_date: str,
    end_date: str,
    reason: str,
) -> dict:
    error = validate_leave(start_date, end_date)
    if error:
        return {"error": error}

    db = get_db()
    leave = {
        "employeeId": employee_id,
        "employeeName": employee_name,
        "startDate": start_date,
        "endDate": end_date,
        "durationDays": days_between(start_date, end_date),
        "reason": reason,
        "status": "pending",
        "createdAt": datetime.utcnow().isoformat(),
    }
    ref = db.collection("leaves").add(leave)
    return {"id": ref[1].id, **leave}


def approve_leave(leave_id: str) -> None:
    db = get_db()
    leave_ref = db.collection("leaves").document(leave_id)
    leave_data = leave_ref.get().to_dict()
    if not leave_data:
        raise ValueError("Leave not found")

    leave_ref.update({"status": "approved"})
    _notify_employee(
        leave_data["employeeId"],
        "leave_approved",
        "בקשת החופשה אושרה",
        f"החופשה שלך מ-{leave_data['startDate']} עד {leave_data['endDate']} אושרה",
    )


def reject_leave(leave_id: str) -> None:
    db = get_db()
    leave_ref = db.collection("leaves").document(leave_id)
    leave_data = leave_ref.get().to_dict()
    if not leave_data:
        raise ValueError("Leave not found")

    leave_ref.update({"status": "rejected"})
    _notify_employee(
        leave_data["employeeId"],
        "leave_rejected",
        "בקשת החופשה נדחתה",
        f"החופשה שלך מ-{leave_data['startDate']} עד {leave_data['endDate']} נדחתה",
    )


def _notify_employee(employee_id: str, notif_type: str, title: str, body: str) -> None:
    db = get_db()
    db.collection("notifications").add({
        "userId": employee_id,
        "type": notif_type,
        "title": title,
        "body": body,
        "read": False,
        "createdAt": datetime.utcnow().isoformat(),
    })

File: src/services/leaveService.ts
import {
  collection,
  addDoc,
  updateDoc,
  doc,
  getDoc,
  query,
  where,
  getDocs,
} from 'firebase/firestore';
import { db } from './firebase';
import { auth } from './firebase';
import { LeaveRequest, LeaveStatus, SHIFT_RULES } from '../types';
import { daysBetween, daysFromNow } from '../utils/dateUtils';

export function validateLeaveRequest(
  startDate: string,
  endDate: string
): string | null {
  const duration = daysBetween(startDate, endDate);
  const daysAhead = daysFromNow(startDate);

  if (duration <= 1) return 'תאריך סיום חייב להיות לפחות יום אחד אחרי תאריך התחלה';

  if (duration <= SHIFT_RULES.SHORT_LEAVE_MAX_DAYS) {
    if (daysAhead < SHIFT_RULES.SHORT_LEAVE_NOTICE_DAYS)
      return `חופשה קצרה דורשת ${SHIFT_RULES.SHORT_LEAVE_NOTICE_DAYS} ימי הודעה מראש`;
  } else {
    if (daysAhead < SHIFT_RULES.LONG_LEAVE_NOTICE_DAYS)
      return `חופשה ארוכה דורשת ${SHIFT_RULES.LONG_LEAVE_NOTICE_DAYS} ימי הודעה מראש`;
  }

  return null;
}

export async function requestLeave(
  employeeId: string,
  employeeName: string,
  startDate: string,
  endDate: string,
  reason: string
): Promise<{ leave: LeaveRequest | null; error: string | null }> {
  if (auth.currentUser && auth.currentUser.uid !== employeeId) {
    return { leave: null, error: 'לא מורשה לבצע פעולה זו' };
  }
  const validationError = validateLeaveRequest(startDate, endDate);
  if (validationError) return { leave: null, error: validationError };

  const leave: Omit<LeaveRequest, 'id'> = {
    employeeId,
    employeeName,
    startDate,
    endDate,
    durationDays: daysBetween(startDate, endDate),
    reason,
    status: 'pending',
    createdAt: new Date().toISOString(),
  };

  const ref = await addDoc(collection(db, 'leaves'), leave);
  return { leave: { ...leave, id: ref.id }, error: null };
}

export async function updateLeaveStatus(
  leaveId: string,
  status: LeaveStatus
): Promise<void> {
  await updateDoc(doc(db, 'leaves', leaveId), { status });

  try {
    const snap = await getDoc(doc(db, 'leaves', leaveId));
    if (!snap.exists()) return;
    const leave = snap.data() as LeaveRequest;
    const title = status === 'approved' ? 'החופשה שלך אושרה ✅' : 'החופשה שלך נדחתה ❌';
    const body = status === 'approved'
      ? `חופשה מ-${leave.startDate} עד ${leave.endDate} אושרה`
      : `חופשה מ-${leave.startDate} עד ${leave.endDate} נדחתה`;
    await addDoc(collection(db, 'notifications'), {
      userId: leave.employeeId,
      title,
      body,
      type: status === 'approved' ? 'leave_approved' : 'leave_rejected',
      read: false,
      createdAt: new Date().toISOString(),
    });
  } catch {
    // notification failure is non-critical
  }
}

export async function getEmployeeLeaves(employeeId: string): Promise<LeaveRequest[]> {
  const q = query(collection(db, 'leaves'), where('employeeId', '==', employeeId));
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as LeaveRequest))
    .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}

export async function getAllPendingLeaves(): Promise<LeaveRequest[]> {
  const q = query(collection(db, 'leaves'), where('status', '==', 'pending'));
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as LeaveRequest))
    .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}
